src/script_opt/ZAM/ZBody.cc

```diff
@@ -0,0 +1,539 @@
+// See the file "COPYING" in the main distribution directory for copyright.
+
+#include "zeek/Desc.h"
+#include "zeek/RE.h"
+#include "zeek/Frame.h"
+#include "zeek/EventHandler.h"
+#include "zeek/Trigger.h"
+#include "zeek/Traverse.h"
+#include "zeek/Reporter.h"
+#include "zeek/script_opt/ScriptOpt.h"
+#include "zeek/script_opt/ZAM/Compile.h"
+
+// Needed for managing the corresponding values.
+#include "zeek/File.h"
+#include "zeek/Func.h"
+#include "zeek/OpaqueVal.h"
+
+// Just needed for BiFs.
+#include "zeek/analyzer/Manager.h"
+#include "zeek/broker/Manager.h"
+#include "zeek/file_analysis/Manager.h"
+#include "zeek/logging/Manager.h"
+
+
+namespace zeek::detail {
+
+using std::vector;
+
+static bool did_init = false;
+
+// Count of how often each type of ZOP executed, and how much CPU it
+// cumulatively took.
+int ZOP_count[OP_NOP+1];
+double ZOP_CPU[OP_NOP+1];
+
+
+void report_ZOP_profile()
+	{
+	for ( int i = 1; i <= OP_NOP; ++i )
+		if ( ZOP_count[i] > 0 )
+			printf("%s\t%d\t%.06f\n", ZOP_name(ZOp(i)),
+			       ZOP_count[i], ZOP_CPU[i]);
+	}
+
+
+// Sets the given element to a copy of an existing (not newly constructed)
+// ZVal, including underlying memory management.  Returns false if the
+// assigned value was missing (which we can only tell for managed types),
+// true otherwise.
+
+static bool copy_vec_elem(VectorVal* vv, int ind, ZVal zv, const TypePtr& t)
+	{
+	if ( vv->Size() <= ind )
+		vv->Resize(ind + 1);
+
+	auto& elem = (*vv->RawVec())[ind];
+
+	if ( ! ZVal::IsManagedType(t) )
+		{
+		elem = zv;
+		return true;
+		}
+
+	if ( elem )
+		ZVal::DeleteManagedType(*elem);
+
+	elem = zv;
+	auto managed_elem = elem->ManagedVal();
+
+	if ( ! managed_elem )
+		{
+		elem = std::nullopt;
+		return false;
+		}
+
+	zeek::Ref(managed_elem);
+	return true;
+	}
+
+// Unary vector operations never work on managed types, so no need
+// to pass in the type ...  However, the RHS, which normally would
+// be const, needs to be non-const so we can use its Type() method
+// to get at a shareable VectorType.
+static void vec_exec(ZOp op, VectorVal*& v1, VectorVal* v2, const ZInst& z);
+
+// Binary operations *can* have managed types (strings).
+static void vec_exec(ZOp op, TypePtr t, VectorVal*& v1, VectorVal* v2,
+                     const VectorVal* v3, const ZInst& z);
+
+// Vector coercion.
+//
+// ### Should check for underflow/overflow.
```

Comment:
Whoops - yes, this needs fixing, and in multiple places.  I'm going to do that as a separate set of commits coming up later, rather than the batch I'm going to push shortly.